CIPs/0017.md

```diff
@@ -0,0 +1,165 @@
+Specify one or more new plumo proofs which have appeared on the network.
+This is intended to be honored with a proceeding GetPlumoProofsMsg to request the proof.
+
+GetPlumoProofsMsg (0x1a)
+[complement: Boolean, [[firstEpoch, lastEpoch, versionNumber], ...]]
+Sent from node A to node B: if `complement` is true, then this message is sent with the set of proofs
+that node A currently *has* and requests those it is missing.
+Node B's reply must contain any proofs that node B has that were not included in the message node A sent.
+If `complement` is false, then this message is sent with the set of proofs that node A currently
+*does not have* and requests these from this peer.
+Node B's reply must contain the proofs requested, if node B has them. Node A should only request proofs
+that node B has indicated via `NewPlumoProof` that it has.
+
+
+PlumoProofMsg (0x1b)
+[plumoProof_0, plumoProof_1, ...]
+Reply to GetPlumoProofMsg. The items in the list are plumo proof structs, {proof: []byte, firstEpoch: uint, lastEpoch: uint, versionNumber: uint}.
+This may validly contain no plumo proofs if the GetPlumoProofMsg included all proofs this node is aware of.
+```
+
+This will be accompanied by a new `proofFetcher` object to be in charge of fetching these proofs from peers while providing protections against DoS attacks. This feels like overkill, but seems necessary to prevent attacks and maintain state.
+
+### `les` protocol changes
+
+Relevant proposed structs:
+```
+// Same as in the `eth` protocol, minimal proof metadata
+ProofMetadata struct {
+  FirstEpoch uint
+  LastEpoch uint
+  VersionNumber uint
+}
+
+// Metadata for each epoch to construct `EpochBlock` for the snark's verify fn
+// May be unnecessary if we can make assumptions on `MaxNonSigners`
+LightEpochBlock struct {
+  Index         uint // Epoch Index 
+  MaxNonSigners uint // MinQuorumSize()
+}
+
+// All of the data needed by the light client to verify and utilize a proof.
+// This is also crafted to allow for batch verifications using `FirstHashToField`
+// My rough calculations show this struct is ~439 bytes for a non-batched proof with no val set differences.
+LightPlumoProof struct {
+  Proof             []byte // Serialized proof. Latest estimate from @psi = 383 bytes
+  FirstEpoch        LightEpochBlock // First epoch this proof is over
+  LastEpoch         LightEpochBlock // Last epoch this proof is over
+  VersionNumber     uint // Proof version number, indicates the corresponding verification key to use
+  LastHashToField  []byte  // Hash to field of Last epoch's `snark.EpochBlock` struct. To be used for batch verification.
+                           // If there is no batch verification, this will be empty. May need an additional boolean to signify this.
+                           // If batch verification should be used, newValidators and deletedValidators will be empty.
+  NewValidators     []blsCrypto.SerializedPublicKey // ValSetDiff(FirstEpoch, LastEpoch), added validators
```

Comment:
We use the `NewValidators` and `DeletedValidators` to determine the final validator set. How is that final validator set ordered, and is the hash of the validator set that is proved by the SNARK dependent on this ordering?

Reply:
Yes, the ordering is important. Maybe there's already a diff message format, but my idea was to begin with a v byte map, where v is the maximum validator set size, each byte of the map tells you what has happened to the validator key that was in that position last time. If the first byte is 0, then that validator key has remained in the same position. If it's 8, then the 9th key from the last validator set is now the first position. If it's 255, then a new public key not present in the last validator set has taken that position. The rest of the diff message is the list of new public keys, in order of appearance. This gives us v bytes + 95n bytes, where n≤v is the number of new validator public keys.

Reply:
When we do validator set diffs for a single epoch transition, i.e. N -> N+1, IIRC the validator set diff implies the order. Given your existing validator set, first you delete all the removed validators, leaving "holes" that are then filled, in order, by the added validators.

However it's not clear to me whether or not that works here

Reply:
Hm good catch, I'll take a look at how the `ValSetDiff` is calculated:
if the way the valset is updated like how you describe, then so long as the `AddedValidators` produced from epoch X to epoch X + 100 maintains the order of each intermediary epoch (I think it should be? If not, wouldn't be that hard to do) then this should be easy to fit into the existing diff format. Of course will think in more detail to verify this

Reply:
It's one thing to have the protocol preserve the order of validators from epoch to epoch (i.e., if validator A is in position 8 in epoch t, then it remains in 8 in epoch t+1 even if the validators in 7 and/or 9 are replaced), but what if A is removed and re-added to the set? We can't also ensure its position is waiting for it. Hence the permutation/ removal bitmap suggestion.